docs/man/man3/SocketEngine.3
.TH "SocketEngine" 3 "19 Dec 2005" "Version 1.0Betareleases" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SocketEngine \- The actual socketengine class presents the same interface on all operating systems, but its private members and internal behaviour should be treated as blackboxed, and vary from system to system and upon the config settings chosen by the server admin.  

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <socketengine.h>\fP
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSocketEngine\fP ()"
.br
.RI "\fIConstructor The constructor transparently initializes the socket engine which the ircd is using. \fP"
.ti -1c
.RI "\fB~SocketEngine\fP ()"
.br
.RI "\fIDestructor The destructor transparently tidies up any resources used by the socket engine. \fP"
.ti -1c
.RI "bool \fBAddFd\fP (int fd, bool readable, char type)"
.br
.RI "\fIAdd a file descriptor to the engine Use AddFd to add a file descriptor to the engine and have the socket engine monitor it. \fP"
.ti -1c
.RI "char \fBGetType\fP (int fd)"
.br
.RI "\fIReturns the type value for this file descriptor This function masks off the X_READBIT value so that the type of the socket can be obtained. \fP"
.ti -1c
.RI "bool \fBDelFd\fP (int fd)"
.br
.RI "\fIDelete a file descriptor f rom the engine This function call deletes a file descriptor from the engine, returning true if it succeeded and false if it failed. \fP"
.ti -1c
.RI "bool \fBWait\fP (std::vector< int > &fdlist)"
.br
.RI "\fIWaits for an event. \fP"
.ti -1c
.RI "\fBstd::string\fP \fBGetName\fP ()"
.br
.RI "\fIReturns the socket engines name This returns the name of the engine for use in /VERSION responses. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::vector< int > \fBfds\fP"
.br
.ti -1c
.RI "int \fBEngineHandle\fP"
.br
.ti -1c
.RI "kevent \fBke_list\fP [65535]"
.br
.ti -1c
.RI "timespec \fBts\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The actual socketengine class presents the same interface on all operating systems, but its private members and internal behaviour should be treated as blackboxed, and vary from system to system and upon the config settings chosen by the server admin. 

The current version supports select, epoll and kqueue.
.PP
Definition at line 66 of file socketengine.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SocketEngine::SocketEngine ()"
.PP
Constructor The constructor transparently initializes the socket engine which the ircd is using. 
.PP
Please note that if there is a catastrophic failure (for example, you try and enable epoll on a 2.4 linux kernel) then this function may bail back to the shell.
.PP
Definition at line 35 of file socketengine.cpp.
.PP
References DEBUG, EngineHandle, and log().
.PP
.nf
36 {
37         log(DEBUG,'SocketEngine::SocketEngine()');
38 #ifdef USE_EPOLL
39         EngineHandle = epoll_create(65535);
40 #endif
41 #ifdef USE_KQUEUE
42         EngineHandle = kqueue();
43 #endif
44 }
.fi
.PP
.SS "SocketEngine::~SocketEngine ()"
.PP
Destructor The destructor transparently tidies up any resources used by the socket engine. 
.PP
Definition at line 46 of file socketengine.cpp.
.PP
References DEBUG, EngineHandle, and log().
.PP
.nf
47 {
48         log(DEBUG,'SocketEngine::~SocketEngine()');
49 #ifdef USE_EPOLL
50         close(EngineHandle);
51 #endif
52 #ifdef USE_KQUEUE
53         close(EngineHandle);
54 #endif
55 }
.fi
.PP
.SH "Member Function Documentation"
.PP 
.SS "bool SocketEngine::AddFd (int fd, bool readable, char type)"
.PP
Add a file descriptor to the engine Use AddFd to add a file descriptor to the engine and have the socket engine monitor it. 
.PP
You must provide a type (see the consts in \fBsocketengine.h\fP) and a boolean flag to indicate wether to watch this fd for read or write events (there is currently no need for support of both).
.PP
Definition at line 65 of file socketengine.cpp.
.PP
References DEBUG, EngineHandle, fds, log(), ref, and X_READBIT.
.PP
Referenced by AddClient(), InspSocket::InspSocket(), and InspSocket::Poll().
.PP
.nf
66 {
67         if ((fd < 0) || (fd > 65535))
68                 return false;
69         this->fds.push_back(fd);
70         ref[fd] = type;
71         if (readable)
72         {
73                 log(DEBUG,'Set readbit');
74                 ref[fd] |= X_READBIT;
75         }
76         log(DEBUG,'Add socket %d',fd);
77 #ifdef USE_EPOLL
78         struct epoll_event ev;
79         log(DEBUG,'epoll: Add socket to events, ep=%d socket=%d',EngineHandle,fd);
80         readable ? ev.events = EPOLLIN | EPOLLET : ev.events = EPOLLOUT | EPOLLET;
81         ev.data.fd = fd;
82         int i = epoll_ctl(EngineHandle, EPOLL_CTL_ADD, fd, &ev);
83         if (i < 0)
84         {
85                 log(DEBUG,'epoll: List insertion failure!');
86                 return false;
87         }
88 #endif
89 #ifdef USE_KQUEUE
90         struct kevent ke;
91         log(DEBUG,'kqueue: Add socket to events, kq=%d socket=%d',EngineHandle,fd);
92         EV_SET(&ke, fd, readable ? EVFILT_READ : EVFILT_WRITE, EV_ADD, 0, 0, NULL);
93         int i = kevent(EngineHandle, &ke, 1, 0, 0, NULL);
94         if (i == -1)
95         {
96                 log(DEBUG,'kqueue: List insertion failure!');
97                 return false;
98         }
99 #endif
100 return true;
101 }
.fi
.PP
.SS "bool SocketEngine::DelFd (int fd)"
.PP
Delete a file descriptor f rom the engine This function call deletes a file descriptor from the engine, returning true if it succeeded and false if it failed. 
.PP
Definition at line 103 of file socketengine.cpp.
.PP
References DEBUG, EngineHandle, fds, log(), ref, and X_READBIT.
.PP
Referenced by kill_link(), kill_link_silent(), InspSocket::Poll(), and Server::UserToPseudo().
.PP
.nf
104 {
105         log(DEBUG,'SocketEngine::DelFd(%d)',fd);
106 
107         if ((fd < 0) || (fd > 65535))
108                 return false;
109 
110         bool found = false;
111         for (std::vector<int>::iterator i = fds.begin(); i != fds.end(); i++)
112         {
113                 if (*i == fd)
114                 {
115                         fds.erase(i);
116                         log(DEBUG,'Deleted fd %d',fd);
117                         found = true;
118                         break;
119                 }
120         }
121 #ifdef USE_KQUEUE
122         struct kevent ke;
123         EV_SET(&ke, fd, ref[fd] & X_READBIT ? EVFILT_READ : EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
124         int i = kevent(EngineHandle, &ke, 1, 0, 0, NULL);
125         if (i == -1)
126         {
127                 log(DEBUG,'kqueue: Failed to remove socket from queue!');
128                 return false;
129         }
130 #endif
131 #ifdef USE_EPOLL
132         struct epoll_event ev;
133         ref[fd] && X_READBIT ? ev.events = EPOLLIN | EPOLLET : ev.events = EPOLLOUT | EPOLLET;
134         ev.data.fd = fd;
135         int i = epoll_ctl(EngineHandle, EPOLL_CTL_DEL, fd, &ev);
136         if (i < 0)
137         {
138                 log(DEBUG,'epoll: List deletion failure!');
139                 return false;
140         }
141 #endif
142         ref[fd] = 0;
143         return found;
144 }
.fi
.PP
.SS "\fBstd::string\fP SocketEngine::GetName ()"
.PP
Returns the socket engines name This returns the name of the engine for use in /VERSION responses. 
.PP
Definition at line 193 of file socketengine.cpp.
.PP
.nf
194 {
195 #ifdef USE_SELECT
196         return 'select';
197 #endif
198 #ifdef USE_KQUEUE
199         return 'kqueue';
200 #endif
201 #ifdef USE_EPOLL
202         return 'epoll';
203 #endif
204         return 'misconfigured';
205 }
.fi
.PP
.SS "char SocketEngine::GetType (int fd)"
.PP
Returns the type value for this file descriptor This function masks off the X_READBIT value so that the type of the socket can be obtained. 
.PP
The core uses this to decide where to dispatch the event to. Please note that some engines such as select() have an upper limit of 1024 descriptors which may be active at any one time, where others such as kqueue have no practical limits at all.
.PP
Definition at line 57 of file socketengine.cpp.
.PP
References ref, and X_EMPTY_SLOT.
.PP
.nf
58 {
59         if ((fd < 0) || (fd > 65535))
60                 return X_EMPTY_SLOT;
61         /* Mask off the top bit used for 'read/write' state */
62         return (ref[fd] & ~0x80);
63 }
.fi
.PP
.SS "bool SocketEngine::Wait (std::vector< int > & fdlist)"
.PP
Waits for an event. 
.PP
Please note that this doesnt wait long, only a couple of milliseconds. It returns a list of active file descriptors in the vector fdlist which the core may then act upon.
.PP
Definition at line 146 of file socketengine.cpp.
.PP
References EngineHandle, fds, ke_list, ref, ts, and X_READBIT.
.PP
.nf
147 {
148         fdlist.clear();
149 #ifdef USE_SELECT
150         FD_ZERO(&wfdset);
151         FD_ZERO(&rfdset);
152         timeval tval;
153         int sresult;
154         for (unsigned int a = 0; a < fds.size(); a++)
155         {
156                 if (ref[fds[a]] & X_READBIT)
157                 {
158                         FD_SET (fds[a], &rfdset);
159                 }
160                 else
161                 {
162                         FD_SET (fds[a], &wfdset);
163                 }
164                 
165         }
166         tval.tv_sec = 0;
167         tval.tv_usec = 100L;
168         sresult = select(FD_SETSIZE, &rfdset, &wfdset, NULL, &tval);
169         if (sresult > 0)
170         {
171                 for (unsigned int a = 0; a < fds.size(); a++)
172                 {
173                         if ((FD_ISSET (fds[a], &rfdset)) || (FD_ISSET (fds[a], &wfdset)))
174                                 fdlist.push_back(fds[a]);
175                 }
176         }
177 #endif
178 #ifdef USE_KQUEUE
179         ts.tv_nsec = 10000L;
180         ts.tv_sec = 0;
181         int i = kevent(EngineHandle, NULL, 0, &ke_list[0], 65535, &ts);
182         for (int j = 0; j < i; j++)
183                 fdlist.push_back(ke_list[j].ident);
184 #endif
185 #ifdef USE_EPOLL
186         int i = epoll_wait(EngineHandle, events, 65535, 100);
187         for (int j = 0; j < i; j++)
188                 fdlist.push_back(events[j].data.fd);
189 #endif
190         return true;
191 }
.fi
.PP
.SH "Member Data Documentation"
.PP 
.SS "int \fBSocketEngine::EngineHandle\fP\fC [private]\fP"
.PP
Definition at line 69 of file socketengine.h.
.PP
Referenced by AddFd(), DelFd(), SocketEngine(), Wait(), and ~SocketEngine().
.SS "std::vector<int> \fBSocketEngine::fds\fP\fC [private]\fP"
.PP
Definition at line 68 of file socketengine.h.
.PP
Referenced by AddFd(), DelFd(), and Wait().
.SS "struct kevent \fBSocketEngine::ke_list\fP[65535]\fC [private]\fP"
.PP
Definition at line 74 of file socketengine.h.
.PP
Referenced by Wait().
.SS "struct timespec \fBSocketEngine::ts\fP\fC [private]\fP"
.PP
Definition at line 75 of file socketengine.h.
.PP
Referenced by Wait().

.SH "Author"
.PP 
Generated automatically by Doxygen for InspIRCd from the source code.


docs/ChangeLog
1.0 Beta 4

    * Fixed configure bug resulting in osflags being called BEFORE getcache (Making it redundant.)
    * Fixed output of src/modules/Makefile saying it was made by Craig.net tongue.gif
    * Updated for test gcc4 compatibility
    * Added support back for ./inspircd
    * Added a line to automatically chmod ./inspircd to 0744 ('Last Second Save!')
    * configure now supports any future .file.inc includes (Yes, i can see into the future)
    * Updated makefile
    * New Perl ./configure
    * Added -Woverloaded-virtual to catch annoying module typos
    * Fixed to always look for the ircd.log within the bin/ dir
    * Fixed to not spool the error log after ./inspircd debug
    * Sanity checks for fd_ref in addclient
    * Fixed 'server ignores me forever if i throttle the server with connections' bug
    * Debugging now allowed from ./inspircd (./inspircd debug)
    * Added LD_LIBRARY_PATH hacks/kludges to the .inspircd.inc file to make it find libmysql in funny situations
    * Added final documentation for inspircd Beta 4
    * Added important note relating to ip addresses in <link:ipaddr>
    * Added schema for m_sqllog.so
    * Added m_sqloper, allows storage of opers within a mysql database
    * Commented new functions
    * Major *MAJOR* optimizations by double-referencing channels to users (never need to scan the entire user hash again except in very rare circumstances)
    * Added channel 'counter', increases speed of quits, parts, kicks
    * commented some users and connections code
    * added remote @* notice
    * Added remote versioning
    * Fixed dicky buffer (due to cleanup the wrong variable was being used)
    * Services speedups for stupid buffered i/o uplinks
    * Added server input buffers (for systems that decide buffering is fun)
    * Major code tidyup (-W)
    * Fixed some flood checking
    * added noticeall
    * Added a Server:: method to send server notices to channels
    * Added explicit oper tracking in a vector of userrec* pointers, optimizes sending out oper notices
    * Fixed: +s and +p channels wouldn't be synched correctly
    * Fixed: +s and +p never show in whois even when you're a member of them
    * Fixed: +s and +p never show in list even if you're a member of them
    * Added Server::FindDescriptor
    * Added and tested fd crossreference table - improves speed a lot, and fixes one of the /STATS chars which has been broken a while
    * Added OnAddBan and OnDelBan module api calls, and fixed glitch which required them in m_timedbans module
    * Added OnOperCompare function to override strcmp in password check for /oper
    * Added module message passing architecture
    * Added a few anti-flood features
    * Updated buffering, faster and more sensible (old code sucks)
    * Added password field
    * Removed some debug output
    * Added /stats T
    * Empty but non-null string passed to is_uline now returns true
    * Added numeric 318 to /WHOIS when nick is not found (end of whois should still be there as its a list response)
    * Added numeric 412 "no text to send"
    * Changed to handle : as a NOP token so that 'assuming' inspircd knows RFC commands wont cause tons of errors
    * Fix to a mode parsing bug (eating wrong version of parameter?)
    * fixed warning of redefinition of RUSAGE_SELF
    * Fix annoying off-by-one error in connection::RecvPacket
    * Added more fault tolerance to server links
    * Added new /stats z
    * Fixes to /who output
    * Added mode 'compression', stops abuse with modes like: +s-s+s-s+s-s
    * Test fixes for /version strings not being set in burst
    * Fixed a reversed strcmp
    * Added 901 numeric - end of modules list
    * Added support for /SQUIT <mask> and checks against /SQUIT *
    * Changed remote kill reason text
    * Fixed strange leading ":" on remote Z/K/G lines
    * Fixed so that quitprefixes are shown in client exit notices
    * Added global oper monitoring through modules and global connect monitoring through modules
    * Fixed weird issues when anope sends large amounts of text at once to the uplink
    * Fixed dodgy poll()
    * Added new and improved GetBuffer function based on research on crappy redhat 7.3
    * Fixed park crash bug #63 reported by ViaraiX
    * Reordered some on-connect network stuff to stop services being so confused
    * Extra safety checks to avoid trying to write() remote users
    * Optimized WriteCommon and WriteCommonExcept
    * Fixed weird line wrapping bug with extremely long lines
    * Made xline notices neater when notifying of unknown connections
    * Fixed WriteCommon not writing to the source user when not on any channels
    * Changed "is now running" message to AFTER port bind checks
    * Added PID reporting and testing before daemonize
    * Changed startup ascii to green. Elphaba would be so proud tongue.gif
    * Fix to prevent empty umodes
    * Fixed empty modes from nickserv
    * Optimized modes (removed strlens etc)
    * Fixed empty modes issue that emerged after dodgy code was corrected
    * Fixed services not being informed of force joins
    * Fix to prevent quits being sent onto the network for nonexistent users
    * Added faster wildcard checking routines
    * Fix for bug #62 (replaces only occur once per line in m_censor)
    * m_chanfilter now works on gcc 2.95.x again.
    * m_chanprotect fixed to remove privilages upon kick
    * Fixed WriteChannelWithServ
    * Added module which provides MD5 encryption for oper passwords plus /mkpasswd
    * Added another check for 'NoServerUline' type thing in m_services.
    * Added Check for 'NoServer' when +r ing a channel in m_services.
    * Added m_timedbans (timed bans module for channel ops/halfops)
    * Added m_sql, module to provide sql db access to other modules
    * Added m_sqlauth - allows authorizing of connections via an arbitary mysql table with a username and password field

src/socketengine.cpp
/*       +------------------------------------+
 *       | Inspire Internet Relay Chat Daemon |
 *       +------------------------------------+
 *
 *  Inspire is copyright (C) 2002-2005 ChatSpike-Dev.
 *                       E-mail:
 *                <brain@chatspike.net>
 *                <Craig@chatspike.net>
 *
 * Written by Craig Edwards, Craig McLure, and others.
 * This program is free but copyrighted software; see
 *            the file COPYING for details.
 *
 * ---------------------------------------------------
 */

#include "inspircd_config.h"
#include "globals.h"
#include "inspircd.h"
#ifdef USE_EPOLL
#include <sys/epoll.h>
#define EP_DELAY 5
#endif
#ifdef USE_KQUEUE
#include <sys/types.h>
#include <sys/event.h>
#include <sys/time.h>
#endif
#include <vector>
#include <string>
#include "socketengine.h"

char ref[65535];

SocketEngine::SocketEngine()
{
	log(DEBUG,"SocketEngine::SocketEngine()");
#ifdef USE_EPOLL
	EngineHandle = epoll_create(65535);
#endif
#ifdef USE_KQUEUE
	EngineHandle = kqueue();
#endif
}

SocketEngine::~SocketEngine()
{
	log(DEBUG,"SocketEngine::~SocketEngine()");
#ifdef USE_EPOLL
	close(EngineHandle);
#endif
#ifdef USE_KQUEUE
	close(EngineHandle);
#endif
}

char SocketEngine::GetType(int fd)
{
	if ((fd < 0) || (fd > 65535))
		return X_EMPTY_SLOT;
	/* Mask off the top bit used for 'read/write' state */
	return (ref[fd] & ~0x80);
}

bool SocketEngine::AddFd(int fd, bool readable, char type)
{
	if ((fd < 0) || (fd > 65535))
		return false;
	this->fds.push_back(fd);
	ref[fd] = type;
	if (readable)
	{
		log(DEBUG,"Set readbit");
		ref[fd] |= X_READBIT;
	}
	log(DEBUG,"Add socket %d",fd);
#ifdef USE_EPOLL
	struct epoll_event ev;
	log(DEBUG,"epoll: Add socket to events, ep=%d socket=%d",EngineHandle,fd);
	readable ? ev.events = EPOLLIN | EPOLLET : ev.events = EPOLLOUT | EPOLLET;
	ev.data.fd = fd;
	int i = epoll_ctl(EngineHandle, EPOLL_CTL_ADD, fd, &ev);
	if (i < 0)
	{
		log(DEBUG,"epoll: List insertion failure!");
		return false;
	}
#endif
#ifdef USE_KQUEUE
	struct kevent ke;
	log(DEBUG,"kqueue: Add socket to events, kq=%d socket=%d",EngineHandle,fd);
	EV_SET(&ke, fd, readable ? EVFILT_READ : EVFILT_WRITE, EV_ADD, 0, 0, NULL);
	int i = kevent(EngineHandle, &ke, 1, 0, 0, NULL);
	if (i == -1)
	{
		log(DEBUG,"kqueue: List insertion failure!");
		return false;
	}
#endif
return true;
}

bool SocketEngine::DelFd(int fd)
{
	log(DEBUG,"SocketEngine::DelFd(%d)",fd);

	if ((fd < 0) || (fd > 65535))
		return false;

	bool found = false;
	for (std::vector<int>::iterator i = fds.begin(); i != fds.end(); i++)
	{
		if (*i == fd)
		{
			fds.erase(i);
			log(DEBUG,"Deleted fd %d",fd);
			found = true;
			break;
		}
	}
#ifdef USE_KQUEUE
	struct kevent ke;
	EV_SET(&ke, fd, ref[fd] & X_READBIT ? EVFILT_READ : EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
	int i = kevent(EngineHandle, &ke, 1, 0, 0, NULL);
	if (i == -1)
	{
		log(DEBUG,"kqueue: Failed to remove socket from queue!");
		return false;
	}
#endif
#ifdef USE_EPOLL
	struct epoll_event ev;
	ref[fd] && X_READBIT ? ev.events = EPOLLIN | EPOLLET : ev.events = EPOLLOUT | EPOLLET;
	ev.data.fd = fd;
	int i = epoll_ctl(EngineHandle, EPOLL_CTL_DEL, fd, &ev);
	if (i < 0)
	{
		log(DEBUG,"epoll: List deletion failure!");
		return false;
	}
#endif
	ref[fd] = 0;
	return found;
}

bool SocketEngine::Wait(std::vector<int> &fdlist)
{
	fdlist.clear();
#ifdef USE_SELECT
	FD_ZERO(&wfdset);
	FD_ZERO(&rfdset);
	timeval tval;
	int sresult;
	for (unsigned int a = 0; a < fds.size(); a++)
	{
		if (ref[fds[a]] & X_READBIT)
		{
			FD_SET (fds[a], &rfdset);
		}
		else
		{
			FD_SET (fds[a], &wfdset);
		}
		
	}
	tval.tv_sec = 0;
	tval.tv_usec = 100L;
	sresult = select(FD_SETSIZE, &rfdset, &wfdset, NULL, &tval);
	if (sresult > 0)
	{
		for (unsigned int a = 0; a < fds.size(); a++)
		{
			if ((FD_ISSET (fds[a], &rfdset)) || (FD_ISSET (fds[a], &wfdset)))
				fdlist.push_back(fds[a]);
		}
	}
#endif
#ifdef USE_KQUEUE
	ts.tv_nsec = 10000L;
	ts.tv_sec = 0;
	int i = kevent(EngineHandle, NULL, 0, &ke_list[0], 65535, &ts);
	for (int j = 0; j < i; j++)
		fdlist.push_back(ke_list[j].ident);
#endif
#ifdef USE_EPOLL
	int i = epoll_wait(EngineHandle, events, 65535, 100);
	for (int j = 0; j < i; j++)
		fdlist.push_back(events[j].data.fd);
#endif
	return true;
}

std::string SocketEngine::GetName()
{
#ifdef USE_SELECT
	return "select";
#endif
#ifdef USE_KQUEUE
	return "kqueue";
#endif
#ifdef USE_EPOLL
	return "epoll";
#endif
	return "misconfigured";
}
